### ISSUE TITLE
[Bug] Why is SDK inference slower than backend inference?

### ISSUE THREAD

**Issue by Y-T-G** [open-mmlab/mmdeploy#2501]
### Checklist

- [x] I have searched related issues but cannot get the expected help.
- [x] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/tree/main/docs/en/faq.md) but cannot get the expected help.
- [x] 3. The bug has not been fixed in the latest version.

### Describe the bug

The SDK inference is much slower than backend inference.

### Reproduction

# SDK Inference
```python
from mmdeploy_runtime import Classifier
import numpy as np

model = Classifier(model_path="./weights", device_name="cuda", device_id=0)

frame = (np.random.rand(224,224,3)*255).astype(np.uint8)

model(frame)
```

`timeit` results for SDK inference: 16.4 ms ± 280 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# Backend inference
```python
deploy_cfg = 'configs/mmpretrain/classification_tensorrt-fp16_dynamic-224x224-224x224.py'
model_cfg = 'mmpretrain/configs/resnet/resnet18_8xb32_in1k.py'
device = 'cuda:0'
backend_model = ['weights/end2end.engine']
image = 'tests/data/tiger.jpeg'

# read deploy_cfg and model_cfg
deploy_cfg, model_cfg = load_config(deploy_cfg, model_cfg)

# build task and backend model
task_processor = build_task_processor(model_cfg, deploy_cfg, device)
model = task_processor.build_backend_model(backend_model)

# process input image
input_shape = get_input_shape(deploy_cfg)
model_inputs, _ = task_processor.create_input(image, input_shape)

# do model inference
with torch.no_grad():
    result = model.test_step(model_inputs)
```

`timeit` results for backend inference: 8.89 ms ± 266 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

Even with a batch of 8, backend inference is faster:
```python
bs = 8
model_inputs, _ = task_processor.create_input(image, input_shape)
inputs = model_inputs['inputs'].copy() 
data_samples = model_inputs['data_samples'].copy() 

for i in range(bs-1):
    model_inputs['inputs'] += inputs
    model_inputs['data_samples'] += data_samples
```
`timeit` results for backend inference: 9.12 ms ± 217 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

### Environment

```Shell
10/17 07:59:20 - mmengine - INFO - 

10/17 07:59:20 - mmengine - INFO - **********Environmental information**********
10/17 07:59:21 - mmengine - INFO - sys.platform: linux
10/17 07:59:21 - mmengine - INFO - Python: 3.8.10 (default, Mar 13 2023, 10:26:41) [GCC 9.4.0]
10/17 07:59:21 - mmengine - INFO - CUDA available: True
10/17 07:59:21 - mmengine - INFO - numpy_random_seed: 2147483648
10/17 07:59:21 - mmengine - INFO - GPU 0: NVIDIA A100 80GB PCIe
10/17 07:59:21 - mmengine - INFO - CUDA_HOME: /usr/local/cuda
10/17 07:59:21 - mmengine - INFO - NVCC: Cuda compilation tools, release 11.3, V11.3.109
10/17 07:59:21 - mmengine - INFO - GCC: x86_64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
10/17 07:59:21 - mmengine - INFO - PyTorch: 1.10.0+cu113
10/17 07:59:21 - mmengine - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

10/17 07:59:21 - mmengine - INFO - TorchVision: 0.11.0+cu113
10/17 07:59:21 - mmengine - INFO - OpenCV: 4.8.0
10/17 07:59:21 - mmengine - INFO - MMEngine: 0.9.0
10/17 07:59:21 - mmengine - INFO - MMCV: 2.0.1
10/17 07:59:21 - mmengine - INFO - MMCV Compiler: GCC 9.3
10/17 07:59:21 - mmengine - INFO - MMCV CUDA Compiler: 11.3
10/17 07:59:21 - mmengine - INFO - MMDeploy: 1.3.0+1090fb6
10/17 07:59:21 - mmengine - INFO - 

10/17 07:59:21 - mmengine - INFO - **********Backend information**********
10/17 07:59:21 - mmengine - INFO - tensorrt:	8.2.3.0
10/17 07:59:21 - mmengine - INFO - tensorrt custom ops:	NotAvailable
10/17 07:59:21 - mmengine - INFO - ONNXRuntime:	1.13.1
10/17 07:59:21 - mmengine - INFO - ONNXRuntime-gpu:	1.8.1
10/17 07:59:21 - mmengine - INFO - ONNXRuntime custom ops:	NotAvailable
10/17 07:59:21 - mmengine - INFO - pplnn:	0.8.1
10/17 07:59:21 - mmengine - INFO - ncnn:	1.0.20230224
10/17 07:59:21 - mmengine - INFO - ncnn custom ops:	NotAvailable
10/17 07:59:21 - mmengine - INFO - snpe:	None
10/17 07:59:21 - mmengine - INFO - openvino:	2022.3.0
10/17 07:59:21 - mmengine - INFO - torchscript:	1.10.0+cu113
10/17 07:59:21 - mmengine - INFO - torchscript custom ops:	NotAvailable
10/17 07:59:21 - mmengine - INFO - rknn-toolkit:	None
10/17 07:59:21 - mmengine - INFO - rknn-toolkit2:	None
10/17 07:59:21 - mmengine - INFO - ascend:	None
10/17 07:59:21 - mmengine - INFO - coreml:	None
10/17 07:59:21 - mmengine - INFO - tvm:	None
10/17 07:59:21 - mmengine - INFO - vacc:	None
10/17 07:59:21 - mmengine - INFO - 

10/17 07:59:21 - mmengine - INFO - **********Codebase information**********
10/17 07:59:21 - mmengine - INFO - mmdet:	3.1.0
10/17 07:59:21 - mmengine - INFO - mmseg:	None
10/17 07:59:21 - mmengine - INFO - mmpretrain:	1.0.2
10/17 07:59:21 - mmengine - INFO - mmocr:	None
10/17 07:59:21 - mmengine - INFO - mmagic:	None
10/17 07:59:21 - mmengine - INFO - mmdet3d:	None
10/17 07:59:21 - mmengine - INFO - mmpose:	1.1.0
10/17 07:59:21 - mmengine - INFO - mmrotate:	None
10/17 07:59:21 - mmengine - INFO - mmaction:	1.1.0
10/17 07:59:21 - mmengine - INFO - mmrazor:	None
10/17 07:59:21 - mmengine - INFO - mmyolo:	None
```

**Comment by RunningLeon** [comment]
hi, you have to exclude the first number of runs for trt backends as warm-up when timing it. Try use https://mmdeploy.readthedocs.io/en/latest/02-how-to-run/useful_tools.html#profiler

**Comment by Y-T-G** [comment]
> hi, you have to exclude the first number of runs for trt backends as warm-up when timing it. Try use https://mmdeploy.readthedocs.io/en/latest/02-how-to-run/useful_tools.html#profiler

Yes. I ran it multiple times, so the warm up shouldn't be included. The profiler uses the API backend, so it is not representative of the speed with the SDK inference.

**Comment by Y-T-G** [comment]
From the profiler

```
----- Settings:
+------------+---------+
| batch size |    1    |
|   shape    | 224x224 |
| iterations |   100   |
|   warmup   |    10   |
+------------+---------+
----- Results:
+--------+------------+---------+
| Stats  | Latency/ms |   FPS   |
+--------+------------+---------+
|  Mean  |   3.893    | 256.859 |
| Median |   3.849    | 259.824 |
|  Min   |   1.839    | 543.765 |
|  Max   |   7.179    | 139.299 |
+--------+------------+---------+
```

From API manually:
`4.31 ms ± 363 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`

From SDK:
`7.44 ms ± 193 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`
